documentation/GuardsOverviewPt1.md

```diff
@@ -0,0 +1,308 @@
+# How TorchDynamo Works - Pt1 - From user code to guards
+
+## Welcome
+Welcome to the first entry in a series of technical documents outlining how TorchDynamo works.
+
+We assume the reader already has a working knowledge of PyTorch, and Python, and a basic understanding of TorchDynamo's role in the PyTorch Ecosystem. The target audience is engineers and researchers looking to learn more about how TorchDynamo operates under the hood.
+
+## Using Dynamo
+From a UX perspective, TorchDynamo is very easy to use. The user invokes `torchdynamo.optimize`, either as a context:
+```py
+with torchdynamo.optimize(my_compiler):
+```
+or as an annotation:
+```py
+@torchdynamo.optimize(my_compiler)
+def fn_foo(bar):
+```
+
+Where a complete example looks like this:
+
+```py
+from typing import List
+import torch
+import torchdynamo
+
+def toy_example(a, b):
+    x = a / (torch.abs(a) + 1)
+    if b.sum() < 0:
+        b = b * -1
+    return x * b
+
+def my_compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
+    print("my_compiler() called with FX graph:")
+    gm.graph.print_tabular()
+    return gm.forward  # return a python callable
+
+with torchdynamo.optimize(my_compiler):
+    for _ in range(100):
+        toy_example(torch.randn(10), torch.randn(10))
+```
+
+This allows TorchDynamo to capture the interpreted Python frames, grab any and all relevant information, and speed things up wherever it can. The speedup comes from a few places, and can be rather dependent on the backend (my_compiler above) provided, but the one speedup we care about most for today's overview is **caching**.
```

Comment:
> the one speedup we care about most for today's overview is caching

It's somewhat unclear about the relationship between caching and speedup. My view is that caching infact makes program slower than eager. Its goal is to avoid recompilation. So we are basically digging a hole, and caching allows us to get out of it. The speedup mostly comes from the backends, and other optimizations.

Reply:
Very good point, will amend it. I was trying to walk a fine line around complexity and readability, but this is a real omission on my end.